Move installation script to main repo

bquan0 opened the pull request:
## Description
I added the `ubuntu.sh` bash install script from the [install_scripts repo](https://github.com/pyne/install_scripts) to the `scripts` folder of this repo. I also edited it slightly to match the commands in `ubuntu_22.04-dev.dockerfile`. 

## Motivation and Context
This change was requested in #1462.

## Changes
Along with adding the install script and updating it to match the dockerfile, I also updated `ubuntu_22.04-dev.dockerfile` to not check for python versions (since we only use version 3.x now) and updated it from ubuntu 20.04 to 22.04, which were two changes requested by @gonuke.

## Behavior
The current behavior is that the install script is in a separate repo and isn't getting tested/updated when the code in this repo is updated. The new behavior should be that the install script is in this repo along with the other code and it will be tested and updated whenever code in this repo is updated. 

## Other Information
There are probably some things yet to be updated/polished in `ubuntu.sh`. When I was editing it, I only updated parts where `ubuntu.sh` had something different or missing compared to the dockerfile. I left the commands that were only in `ubuntu.sh` but not in the dockerfile intact . 
Also, I'm not sure how to go about testing this bash install script.

bquan0 committed:
remove checks for py_version being 3.x

bquan0 committed:
initial edits to bash script

bquan0 committed:
finish modifying bash script

bquan0 committed:
try using build arg in base ubuntu image name

bquan0 committed:
update docker_publish.yml to use 22.04 instead of 20.04

bquan0 committed:
update CHANGELOG.rst

bquan0 committed:
replace PY_SUFIX with 3 and replace 20.04 with 22.04

ngranda commented on docker/ubuntu_22.04-dev.dockerfile:
```diff
@@ -1,22 +1,19 @@
 ARG py_version=3.8
```
Is there a reason why Python 3.8 and not 3.10 which is the default on Ubuntu 22.04?

gonuke replied:
Good catch @ngranda - I think getting rid of this build arg entirely probably makes sense

bquan0 committed:
remove py_version arg and replace it with 3.10

gonuke commented on docker/ubuntu_22.04-dev.dockerfile:
```diff
@@ -105,7 +96,7 @@ RUN export PYMOAB_FLAG="-DENABLE_PYMOAB=ON"; \
 ENV LD_LIBRARY_PATH $HOME/opt/moab/lib:$LD_LIBRARY_PATH
 ENV LIBRARY_PATH $HOME/opt/moab/lib:$LIBRARY_PATH
-ENV PYTHONPATH=$HOME/opt/moab/lib/python${py_version}/site-packages/
+ENV PYTHONPATH=$HOME/opt/moab/lib/python3.10/site-packages/
```
This isn't robust to the introduction of newer versions of python, but I'm not 100% sure what the solution is.

gonuke reviewed (changes requested):
Thanks @bquan0 .

Can you also add a workflow that runs the install script on a clean Ubuntu 22.04 runner? i.e. not one of our docker images

bquan0 committed:
add install_script.yml

gonuke commented:
Thanks for adding the new workflow but I don’t see it being triggered anywhere

bquan0 committed:
edit install_script.yml to trigger workflow run

bquan0 committed:
make install_script.yml run on every push

bquan0 committed:
add checkout repo step

bquan0 committed:
make install_script.yml run when it's changed, not docker_publish.yml

bquan0 committed:
update test script name in docker_publish.yml

bquan0 committed:
install pytest in dockerfile

bquan0 committed:
use value instead of PYMOAB_FLAG variable

bquan0 committed:
edit PYTHONPATH

bquan0 committed:
change 3.8 to 3.10 in docker_publish.yml

bquan0 committed:
match PYTHONPATH between workflow and dockerfile

bquan0 committed:
change PATH

bquan0 commented:
I noticed that in the `Building PyNE` step of the `BuildTest` job, the `PYTHONPATH` was specified with python3.8 instead of python3.10. I tried changing it to see if it would fix the `ModuleNotFoundError: No module named 'pymoab'` errors I've been getting on the `BuildTest` parts of the [workflow run](https://github.com/bquan0/pyne/actions/runs/4769047799), but it didn't work.

bquan0 committed:
revert changes to PATH and PYTHONPATH

pshriwise committed:
Install PyMOAB to Python system environment

pshriwise committed:
Install PyMOAB to Python environment as user

bquan0 committed:
Merge pull request #1 from pshriwise/pymoab-pip-install

Install PyMOAB to Python environment as user

bquan0 committed:
add backslash

pshriwise commented:
Regarding our attempt at installing PyMOAB to the user location today: I found that the reason this doesn't work is because the moab & pymoab installation happens during the Dockerfile build when `HOME=/root`. Later, when the action uses the container `HOME=/github/home` and the pymoab installation we did can't be found in `$HOME/.local`, one of the default search paths for Python packages. 

I'm not sure what the right fix for this is. There's an issue on it [here](https://github.com/actions/runner/issues/863) without any clear solution that I could see.

FWIW I diagnosed this using https://github.com/marketplace/actions/debugging-with-tmate, a really handy tool that lets you ssh into the CI machine at any point in the build.

gonuke commented:
> Regarding our attempt at installing PyMOAB to the user location today: I found that the reason this doesn't work is because the moab & pymoab installation happens during the Dockerfile build when `HOME=/root`. Later, when the action uses the container `HOME=/github/home` and the pymoab installation we did can't be found in `$HOME/.local`, one of the default search paths for Python packages.
> 
> I'm not sure what the right fix for this is. There's an issue on it [here](https://github.com/actions/runner/issues/863) without any clear solution that I could see.
> 
> FWIW I diagnosed this using https://github.com/marketplace/actions/debugging-with-tmate, a really handy tool that lets you ssh into the CI machine at any point in the build.

This is only a problem because we are installing with `--user` right?  I guess for CI we don't need to do that?

pshriwise commented:
True. We could do a system install instead. I'll try that.

gonuke commented:
We could also move all testing to be part of a docker build like we are aiming towards for DAGMC

pshriwise commented:
Also true. I'll try this first to confirm the plan for the PyMOAB pip install is viable.

pshriwise commented:
Huzzah! https://github.com/pshriwise/pyne/actions/runs/4840438166

pshriwise commented:
I'll submit another PR into @bquan0's branch after a little history cleanup

pshriwise committed:
Merge branch 'install-script' into pymoab-pip-install

bquan0 committed:
Merge pull request #2 from pshriwise/pymoab-pip-install

PyMOAB pip install

gonuke commented on scripts/ubuntu.sh:
```diff
@@ -0,0 +1,220 @@
+          -DBUILD_SHARED_LIBS=ON \
+          -DENABLE_BLASLAPACK=OFF \
+          -DENABLE_FORTRAN=OFF    
+make -j 3 
```
I think we should add 
```
cd pymoab && pip3 install .
```
like we did in the Dockerfile

gonuke commented on scripts/ubuntu.sh:
```diff
@@ -0,0 +1,220 @@
+
+# Run tests
+cd ${install_dir}/pyne/pyne-repo/tests
+./travis-run-tests.sh
```
Isn't this 
```suggestion
./ci-run-tests.sh
```
Now?

gonuke commented on docker/ubuntu_22.04-dev.dockerfile:
```diff
@@ -98,14 +90,17 @@ RUN export PYMOAB_FLAG="-DENABLE_PYMOAB=ON"; \
 ENV LD_LIBRARY_PATH $HOME/opt/moab/lib:$LD_LIBRARY_PATH
 ENV LIBRARY_PATH $HOME/opt/moab/lib:$LIBRARY_PATH
-ENV PYTHONPATH=$HOME/opt/moab/lib/python${py_version}/site-packages/
+ENV PYTHONPATH=$HOME/opt/moab/lib/python3.10/site-packages/
```
Might not need this anymore if we install pymoab using pip

gonuke reviewed (changes requested):
A few more changes related to pip installing moab

bquan0 committed:
change travis to ci in ubuntu.sh

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

bquan0 committed:
add pip install . tubuntu.sh andand comment out PYTHONPATH ENV var

bquan0 committed:
check what server-tag outputs are from multistage action

gonuke commented on scripts/ubuntu.sh:
```diff
@@ -0,0 +1,223 @@
+          -DENABLE_FORTRAN=OFF    
+make -j 3 
+cd pymoab
+pip install .
```
We need to be careful about how users will run this, i.e. not as `root`.  Should install this as `--user` and set `PYTHONPATH` accordingly

bquan0 replied:
Should I apply this change for all instances of `pip install .`?

gonuke replied:
Probably makes sense to do them all since that's what a user may do?

gonuke commented on scripts/ubuntu.sh:
```diff
@@ -0,0 +1,223 @@
+cd ${install_dir}
+git clone https://github.com/openmc-dev/openmc.git
+cd openmc
+git checkout tags/v0.13.0
```
Can we clone only this tag, as we do with Moab and DAGMC above?

gonuke commented on scripts/ubuntu.sh:
```diff
@@ -0,0 +1,223 @@
+make
+make install
+cd ..
+pip3 install .
```
I think we've aliased `pip` to `pip3` already and used just `pip` above. Let's be consistent

gonuke commented on scripts/ubuntu.sh:
```diff
@@ -0,0 +1,223 @@
+cd pyne-repo
+if [ $1 == 'stable' ] ; then
+  TAG=$(git describe --abbrev=0 --tags)
+  git checkout tags/`echo ${TAG}` -b `echo ${TAG}`
```
Can we do this if we've only cloned a single branch?

bquan0 replied:
I don't think lines 176-178 are being run. In [this workflow run](https://github.com/bquan0/pyne/actions/runs/4960351409/jobs/8875715298), line 1034 corresponds to the `git clone` on line 173 in `ubuntu.sh`, and the lines that follow in the workflow run don't seem to have any tags echoed.

gonuke replied:
I think they are being run but the condition is never true.  If it were ever true, it would probably fail.  Since this is a sample script for users to use to install, we should choose a good practice.

I think this was to allow people to run this script to install a guaranteed stable version rather than the develop branch.  That's probably a good thing to preserve, and therefore we might want to test both versions

bquan0 replied:
After implementing the matrix idea for testing both the develop branch and the stable branch, it seems like it is possible to switch to the stable branch even though only a single branch was cloned. In line 924 of [this workflow run](https://github.com/bquan0/pyne/actions/runs/5150609113/jobs/9274941733) with a command line arg of `stable`, it says `Switched to a new branch '0.7.7'`. This line is not there in [the workflow run](https://github.com/bquan0/pyne/actions/runs/5150609113/jobs/9274941603) with a different command line arg (see line 929)

gonuke reviewed (changes requested):
A few suggestions for users running not as root.  I wonder if we should update the workflow that tests the install script to run as non-root?

bquan0 committed:
change pip3 into pip, clone only one branch

bquan0 committed:
add --user for pip install

bquan0 committed:
run install script with and without if statement

bquan0 committed:
fix errors in install_script.yml

bquan0 committed:
fix indentation

bquan0 commented:
After looking at the workflow runs, it seems like the install script has been running into errors such as `./ubuntu.sh: line 191: nuc_data_make: command not found` and `./ci-run-tests.sh: line 11: pytest: command not found` since [the very first successful workflow run](https://github.com/bquan0/pyne/actions/runs/4759177570/jobs/8458152347), which probably prevent PyNE from being installed (and is also probably why the workflow runs only take ~5 mins)

The new workflow runs run into an error of not being able to find the moab repo (line 302 of [this workflow run](https://github.com/bquan0/pyne/actions/runs/5150609113/jobs/9274941603))

gonuke commented on scripts/ubuntu.sh:
```diff
@@ -0,0 +1,222 @@
+                  cython \
+                  nose \
+                  tables \
+                  matplotlib \
```
need to add `pytest` here - see the Dockerfile for guidance

gonuke commented on docker/ubuntu_22.04-dev.dockerfile:
```diff
@@ -69,15 +62,14 @@ ENV LD_LIBRARY_PATH $HDF5_INSTALL_PATH/lib:$LD_LIBRARY_PATH
 ENV INSTALL_PATH=$HOME/opt/moab
 
 # build MOAB
 RUN export PYMOAB_FLAG="-DENABLE_PYMOAB=ON"; \
```
If we hardcode this flag to be ON, then why not just remove the use of a variable and hard code this in the cmake command below?

gonuke commented on docker/ubuntu_22.04-dev.dockerfile:
```diff
@@ -98,14 +90,17 @@ RUN export PYMOAB_FLAG="-DENABLE_PYMOAB=ON"; \
 ENV LD_LIBRARY_PATH $HOME/opt/moab/lib:$LD_LIBRARY_PATH
 ENV LIBRARY_PATH $HOME/opt/moab/lib:$LIBRARY_PATH
-ENV PYTHONPATH=$HOME/opt/moab/lib/python${py_version}/site-packages/
+# ENV PYTHONPATH=$HOME/opt/moab/lib/python3.10/site-packages/
```
remove this if we don't need it anymore, rather than comment

gonuke commented on scripts/ubuntu.sh:
```diff
@@ -0,0 +1,222 @@
+# Adding MOAB/lib to $LD_LIBRARY_PATH and $LIBRARY_PATH
+export LD_LIBRARY_PATH="${install_dir}/moab/lib:$LD_LIBRARY_PATH"
+
+# Adding pymoab to $PYTHONPATH
+PYTHON_VERSION=$(python -c 'import sys; print(sys.version.split('')[0][0:3])')
+if [ -z $PYTHONPATH ]; then
+  export PYTHONPATH="$install_dir/moab/lib/python${PYTHON_VERSION}/site-packages"
+else
+  export PYTHONPATH="$install_dir/moab/lib/python${PYTHON_VERSION}/site-packages:$PYTHONPATH"
+fi
```
Might not need this if we install with `pip` - see how this works in the dockerfile?

gonuke reviewed (changes requested):
I noted a few things, but most are related to using the working(?) Dockerfile to guide the updates of the install script

gonuke commented:
> this workflow run

definitely strange that it could not find the repo - I wonder if that was a transient error that would be fixed by running the workflow again?

bquan0 committed:
add suggested changes

bquan0 committed:
delete adding pymoab to PYTHONPATH

gonuke commented:
It looks like this is still not successful.  I recommend that you do a line-by-line comparison of the Dockerfile with the ubuntu.sh script.  We know that the former is working, so it is a good guide for creating this script.

The script does historically have the goal of being a kind of tutorial for building pyne as well, but it might be useful to first just get it working and then we can worry about how to embellish for tutorial purposes.

bquan0 committed:
make ubuntu.sh match dockerfile

bquan0 committed:
reference matrix variable build_hdf5 properly

bquan0 commented:
I think after I added the if statements for hdf5 that were in the dockerfile but not `ubuntu.sh`, it seems to be working. I also added another matrix variable to make the script build with hdf5 and without hdf5. The [workflow run](https://github.com/bquan0/pyne/actions/runs/5193706150/jobs/9364586964) lasts longer now and I think the previous errors I mentioned have been fixed.

gonuke commented:
Definitely much better!!  It's building and installing everything correctly, but it doesn't find the script to run the tests.  It's the second last line in the CI log for the "run install script" job. I'm not quite sure why, except that you may be in the wrong directory? Or the file is not executable?

bquan0 commented:
The tests run in the workflows where we use the `develop` branch instead of the stable branch. I ran the git commands for pyne (lines 187 to 192 in `ubuntu.sh`) and found that when the code in the stable if statement ran, we got switched to the branch `0.7.7`, which has a file called `travis-run-tests.sh` instead of the file `ci-run-tests.sh` in the `tests` directory. I think this got changed with [Nancy's PR](https://github.com/pyne/pyne/pull/1478#event-9007835711).

For now, I'm just going to use an if statement to make it run `./travis-run-tests.sh` if the stable branch is used and `./ci-run-tests.sh` if the develop branch is used. After a new stable version is released with the new test script name, we can omit this part  and just do `./ci-run-tests.sh`. 

It looks like the tests are now running on [the workflows that use the stable branch](https://github.com/bquan0/pyne/actions/runs/5213933014/jobs/9409572807), but they fail at `test_ace.py`. This is probably due to the fact that we download python 3.10 in the install script and the old tests in `travis-run-test.sh` are designed for an older version of python. 

After a quick discussion at the software meeting on 6/9, it seems like our options are to either
- stop supporting the stable branch and only worry about the develop branch
- use if statements to download an older version of python when we run it on the stable branch.

bquan0 committed:
add if statement to make tests run in stable branch

gonuke commented:
Good catch on the conditional for stable vs develop.  I think this is all working now so I will merge.

We'll probably want to do a release and update stable (and then update this again).
